x/superfluid/keeper/hooks.go

```diff
@@ -10,50 +11,73 @@ import (
-				if err != nil {
-					k.Logger(ctx).Error(err.Error())
-					k.SetEpochOsmoEquivalentTWAP(ctx, epochNumber, asset.Denom, sdk.NewDec(0))
-					continue
-				}
-
-				// get OSMO amount
-				bondDenom := k.sk.BondDenom(ctx)
-				osmoPoolAsset, err := pool.GetPoolAsset(bondDenom)
-				if err != nil {
-					k.Logger(ctx).Error(err.Error())
-					k.SetEpochOsmoEquivalentTWAP(ctx, epochNumber, asset.Denom, sdk.NewDec(0))
-					continue
-				}
-
-				twap := osmoPoolAsset.Token.Amount.ToDec().Quo(pool.GetTotalShares().Amount.ToDec())
-				k.SetEpochOsmoEquivalentTWAP(ctx, epochNumber, asset.Denom, twap)
-			} else if asset.AssetType == types.SuperfluidAssetTypeNative {
-				// TODO: should get twap price from gamm module and use the price
-				// which pool should it use to calculate native token price?
-				k.Logger(ctx).Error("unsupported superfluid asset type")
-			}
-		}
+		// cref [#830](https://github.com/osmosis-labs/osmosis/issues/830),
+		// the supplied epoch number is wrong at time of commit. hence we get from the info.
+		endedEpochNumber := k.ek.GetEpochInfo(ctx, epochIdentifier).CurrentEpoch
 
 		// Move delegation rewards to perpetual gauge
 		k.MoveSuperfluidDelegationRewardToGauges(ctx)
 
-		// Refresh intermediary accounts' delegation amounts
+		// Update all LP tokens TWAP's for the upcoming epoch.
+		// This affects staking reward distribution until the next epochs rewards.
+		// Exclusive of current epoch's rewards, inclusive of next epoch's rewards.
+		for _, asset := range k.GetAllSuperfluidAssets(ctx) {
+			err := k.updateEpochTwap(ctx, asset, endedEpochNumber)
+			if err != nil {
+				// TODO: Revisit what we do here. (halt all distr, only skip this asset)
+				// Since at MVP of feature, we only have one pool of superfluid staking,
+				// we can punt this question.
+				// each of the errors feels like significant misconfig
+				return
+			}
+		}
+
+		// Refresh intermediary accounts' delegation amounts,
+		// making staking rewards follow the updated TWAP numbers.
 		k.RefreshIntermediaryDelegationAmounts(ctx)
 	}
 }
 
+func (k Keeper) updateEpochTwap(ctx sdk.Context, asset types.SuperfluidAsset, endedEpochNumber int64) error {
+	if asset.AssetType == types.SuperfluidAssetTypeLPShare {
+		// LP_token_Osmo_equivalent = OSMO_amount_on_pool / LP_token_supply
+		poolId := gammtypes.MustGetPoolIdFromShareDenom(asset.Denom)
+		pool, err := k.gk.GetPool(ctx, poolId)
+		if err != nil {
+			// Pool has been unexpectedly deleted
+			k.Logger(ctx).Error(err.Error())
+			k.BeginUnwindSuperfluidAsset(ctx, 0, asset)
+			return err
+		}
+
+		// get OSMO amount
+		bondDenom := k.sk.BondDenom(ctx)
+		osmoPoolAsset, err := pool.GetPoolAsset(bondDenom)
+		if err != nil {
+			// Pool has unexpectedly removed Osmo from its assets.
+			k.Logger(ctx).Error(err.Error())
+			k.BeginUnwindSuperfluidAsset(ctx, 0, asset)
+			return err
+		}
+
+		twap := osmoPoolAsset.Token.Amount.ToDec().Quo(pool.GetTotalShares().Amount.ToDec())
```

Comment:
Maybe we should add a comment here indicating that we're faking TWAP with spot price.

Reply:
Agreed

Reply:
Resolved in latest commit